Update chapter 12 (#218)
hjwp/Book-TDD-Web-Dev-Python: chapter_12_organising_test_files.asciidoc

Brilliant--no need to sit around waiting for all the FTs
when we're only interested in a single one.

> Although we need to remember to run all of them now and again, to check for regressions.
> It's a good habit to run all the tests before making a `git commit`.
> Later in the book we'll see how to give that task over to an automated Continuous Integration loop.
> For now, let's commit!

Reviewer: i'm worried this ends up being ambiguous - are we giving the task of making commits over to CI?

```suggestion
Although we need to remember to run all of them now and again, to check for regressions.
Later in the book we'll set up a Continuous Integration (CI) server to run all the tests automatically,
for example every time we push to master.
For now, a good prompt for running all the tests is "just before you do a commit",
so let's get into that habit now:
```

Author: Nice, I like it!

Reviewer: maybe we'd want to mark it as xfail then?

((("Python 3", "lambda functions")))
`lambda` in Python is the syntax for making a one-line, throwaway function--it

> saves you from having to use `def..():` and an indented block:
> // CSANAD: the `def..` and the `():` renders apart from each other, on two
> //         lines. I couldn't find out how to keep them together.

Reviewer: there's a special role thingie called `keep-together` - syntax is something like this,
but it might not render quite right if we use it at the same time as backticks:

```suggestion
saves you from having to use [keep-together]#`def..():`# and an indented block:.
```

could also try just `[keep-together]`def..():` hmmm

Author: Yeah, I tried that but didn't work. How about cheating a little with an extra word or two:
 "...from having to use **the regular** `def..():` and an indented block:" ?

Reviewer: just ignore it for now, they'll pick it up in copyediting.

((("Python 3", "lambda functions")))
`lambda` in Python is the syntax for making a one-line, throwaway function--it

> saves you from having to use `def..():` and an indented block:

Author: Apparently I can't make my comment below a committable change, so I'll just suggest it again.
I tried the `keep-together` but it didn't work within a code block. How about cheating by adding a word:
```
saves you from having to use the regular `def..():` and an indented block:
```

5
----

> In our case, we're using it to transform a bit of code - that would otherwise be
> executed immediately - into a function that we can pass as an argument,

Reviewer: ```suggestion
In our case, we're using it to transform a bit of code--that would otherwise be
executed immediately--into a function that we can pass as an argument,
```

and that can be executed later, and multiple times:

> // CSANAD: I think it's easier to read like this.

Reviewer: not sure i agree but nbd.

fwiw the preferred syntax for emdashes is `word--nextword`

The order of the traceback is a little confusing, but we can more or less follow
through what happened:

<1> In our FT, we call our `self.wait_for` helper, where we pass
    the `lambda`-ified version of `assertEqual`.

> <2> We go into `self.wait_for` in `base.py`,

Reviewer: oreilly convention is that filenames are in italics 

```suggestion
<2> We go into `self.wait_for` in _base.py_,
```

    where we're calling (and returning) `fn()`, which refers to the passed
    lambda function encapsulating our test assertion.

<3> To explain where the exception has actually come from,
    the traceback takes us back into _test_list_item_validation.py_
    and inside the body of the `lambda` function,
    and tells us that the attempt to find the `.invalid-feedback` element
    that failed.
// CSANAD: I changed two things here:
//   First, I think it's better to remove the specific line number which may
// vary due to personal formatting preferences (and following the pep8 puts

> // this on line 16 for me)

Reviewer: 💯
